### PR TITLE
Detect Inno, NSIS, and Burn automagically

### LINKED ISSUE #177: [New Feature]: Identify Inno and Nullsoft installers automatically
### Description of the new feature/enhancement

Automatically identify Inno and Nullsoft installers in YamlCreate

### Proposed technical implementation details (optional)

The first **264** bytes of installers created with Inno setup are always the same:

```kotlin
77, 90, 80, 0, 2, 0, 0, 0, 4, 0, 15, 0, -1, -1, 0, 0, -72, 0, 0, 0, 0, 0, 0, 0, 64, 0, 26, 0, 0, 0, 0, 0, 0, 0,
0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, -70, 16, 0, 14, 31,
-76, 9, -51, 33, -72, 1, 76, -51, 33, -112, -112, 84, 104, 105, 115, 32, 112, 114, 111, 103, 114, 97, 109, 32,
109, 117, 115, 116, 32, 98, 101, 32, 114, 117, 110, 32, 117, 110, 100, 101, 114, 32, 87, 105, 110, 51, 50, 13,
10, 36, 55, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 80, 69, 0, 0, 76, 1, 10,
0
```

If you compare the first 264 bytes of an exe to that, you can know whether it has been created with Inno Setup or not.

This is the same with installer created with NSIS, where the first **224** bytes are the same:

```kotlin
77, 90, -112, 0, 3, 0, 0, 0, 4, 0, 0, 0, -1, -1, 0, 0, -72, 0, 0, 0, 0, 0, 0, 0, 64, 0, 0, 0, 0, 0, 0, 0, 0, 0,
0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, -40, 0, 0, 0, 14, 31, -70, 14, 0,
-76, 9, -51, 33, -72, 1, 76, -51, 33, 84, 104, 105, 115, 32, 112, 114, 111, 103, 114, 97, 109, 32, 99, 97, 110,
110, 111, 116, 32, 98, 101, 32, 114, 117, 110, 32, 105, 110, 32, 68, 79, 83, 32, 109, 111, 100, 101, 46, 13, 13,
10, 36, 0, 0, 0, 0, 0, 0, 0, -83, 49, 8, -127, -23, 80, 102, -46, -23, 80, 102, -46, -23, 80, 102, -46, 42, 95,
57, -46, -21, 80, 102, -46, -23, 80, 103, -46, 76, 80, 102, -46, 42, 95, 59, -46, -26, 80, 102, -46, -67, 115,
86, -46, -29, 80, 102, -46, 46, 86, 96, -46, -24, 80, 102, -46, 82, 105, 99, 104, -23, 80, 102, -46, 0, 0, 0, 0,
0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 80, 69, 0, 0, 76, 1, 5, 0
```

### LINKED ISSUE #179: [New Feature]: Identify Burn installers automatically
### Description of the new feature/enhancement

Burn installers can't be identified by magic bytes like #177 but they can be identified by the prescence of a `.wixburn` header.

The only exception to this I've found is JVM apps packaged with JPackager don't contain the `.wixburn` header, which I'm pretty sure is a bug.

### Proposed technical implementation details (optional)

See https://github.com/AnalogJ/Wix3.6Toolset/blob/master/RC0-source/wix36-sources/src/wix/BurnCommon.cs#L252 for Wix's implementation.

Komac's implementation (using [Okio](https://github.com/square/okio)):
```kotlin
val FileHandle.isBurn: Boolean
    get() {
        source().buffer().use { bufferedSource ->
            val sink = Buffer()
            var offset = 0L
            val wixBurnHeaderBytes = wixBurnHeader.encodeUtf8()
            repeat(UShort.MAX_VALUE.toInt()) {
                bufferedSource.read(sink, burnBufferSize)
                if (sink.rangeEquals(offset, wixBurnHeaderBytes)) return true
                offset += burnBufferSize
            }
            return false
        }
    }
    
const val wixBurnHeader = ".wixburn"
```

### PR DESCRIPTION
**Pull request by Trenly**
- [x] Have you signed the [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/winget-pkgs)?

* Resolves Trenly/winget-pkgs#177
* Resolves Trenly/winget-pkgs#179

cc @denelon @russelbanks @mdanish-kh
 ###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-pkgs/pull/106356)

### CODE REVIEW

**Comment by russellbanks** [comment]
If you're always reading 264 bytes, won't NSIS installers always equal false as their magic bytes are 224 bytes long? You'd end up with 40 bytes extra that are different between NSIS installers

**Comment by Trenly** [comment]
> If you're always reading 264 bytes, won't NSIS installers always equal false as their magic bytes are 224 bytes long? You'd end up with 40 bytes extra that are different between NSIS installers

You right; I just didn't read carefully

**Review by russellbanks — DISMISSED** [review]

**Review by vedantmgoyal9 — COMMENTED** [review]
`$burn = @(46; 119; 105; 120; 98; 117; 114; 110)`

The only thing is that it can present anywhere in the binary, so we have to read the complete binary from beginning to end.

**Comment by Trenly** [comment]
> `$burn = @(46; 119; 105; 120; 98; 117; 114; 110)`
> 
> The only thing is that it can present anywhere in the binary, so we have to read the complete binary from beginning to end.

That could present a lot of false positives

**Comment by vedantmgoyal2009** [comment]
It is short because it's just `.wixburn` header as array of UTF-8 bytes.

**Comment by russellbanks** [comment]
> It is short because it's just `.wixburn` header (reference: [Trenly#179](https://github.com/Trenly/winget-pkgs/issues/179)) as an array of UTF-8 bytes.
> 
> The thing is that @russellbanks implemented a different way of detecting `.wixburn` header in binary, but just out of curiosity, I had put these bytes into the same function which detects inno/nullsoft installers with the only modification of reading the complete binary, and it turned out to be working for burn too.

Komac keeps reading a small amount of bytes to find the `.wixburn` header until a certain point. It was based on Wix's C# implementation. Reading the whole binary at once will of course work but will use very large amounts of memory.

For example, say the header can appear anywhere from bytes 100-1000 in a 10,000 byte file, I keep reading 8 bytes at a time until I find it in the 100-1000 section, to keep memory usage at a minimum.

**Comment by vedantmgoyal2009** [comment]
> Reading the whole binary at once will of course work but will use very large amounts of memory.

That's really a disadvantage and since YamlCreate is meant to be a user-facing script, we might have to go through the process of reading a small amount of bytes at a time.

**Review by vedantmgoyal9 — DISMISSED** [review]

**Review by jedieaston — DISMISSED** [review]
...talk about magic numbers :)

**Review by russellbanks — DISMISSED** [review]

File: Tools/YamlCreate.ps1 [hunk 1]
```diff
@@ -1205,11 +1282,11 @@ Function Read-QuickInstallerEntry {
         # Here we also want to pass any exceptions through for potential debugging
         throw [System.Net.WebException]::new('The file could not be downloaded. Try running the script again', $_.Exception)
       }
-      # Check that MSI's aren't actually WIX
+      # Check that MSI's aren't actually WIX, and EXE's aren't NSIS or INNO
```
Comment (mdanish-kh) [inline Tools/YamlCreate.ps1:1285]:
```suggestion
      # Check that MSI's aren't actually WIX, and EXE's aren't NSIS, INNO or BURN
```

File: Tools/YamlCreate.ps1 [hunk 2]
```diff
@@ -2703,17 +2780,17 @@ Switch ($script:Option) {
...
+      $_Installer['InstallerUrl'] = [System.Web.HttpUtility]::UrlDecode($_Installer.InstallerUrl.Replace('+', '%2B'))
       try {
         $script:dest = Get-InstallerFile -URI $_Installer.InstallerUrl -PackageIdentifier $PackageIdentifier -PackageVersion $PackageVersion
       } catch {
         # Here we also want to pass any exceptions through for potential debugging
         throw [System.Net.WebException]::new('The file could not be downloaded. Try running the script again', $_.Exception)
       }
-      # Check that MSI's aren't actually WIX
-      if ($_Installer['InstallerType'] -eq 'msi') {
+      # Check that MSI's aren't actually WIX, and EXE's aren't NSIS or INNO
```
Comment (mdanish-kh) [inline Tools/YamlCreate.ps1:2790]:
```suggestion
      # Check that MSI's aren't actually WIX, and EXE's aren't NSIS, INNO or BURN
```

File: Tools/YamlCreate.ps1 [hunk 3]
```diff
@@ -2703,17 +2780,17 @@ Switch ($script:Option) {
...
+      if ($_Installer['InstallerType'] -in @('msi'; 'exe')) {
         $DetectedType = Get-PathInstallerType $script:dest
-        if ($DetectedType -in @('msi'; 'wix')) { $_Installer['InstallerType'] = $DetectedType }
+        if ($DetectedType -in @('msi'; 'wix'; 'nullsoft'; 'inno')) { $_Installer['InstallerType'] = $DetectedType }
```
Comment (mdanish-kh) [inline Tools/YamlCreate.ps1:2793]:
```suggestion
        if ($DetectedType -in @('msi'; 'wix'; 'nullsoft'; 'inno'; 'burn')) { $_Installer['InstallerType'] = $DetectedType }
```

File: Tools/YamlCreate.ps1 [hunk 4]
```diff
@@ -1205,11 +1282,11 @@ Function Read-QuickInstallerEntry {
...
         throw [System.Net.WebException]::new('The file could not be downloaded. Try running the script again', $_.Exception)
       }
-      # Check that MSI's aren't actually WIX
+      # Check that MSI's aren't actually WIX, and EXE's aren't NSIS or INNO
       Write-Host -ForegroundColor 'Green' "Installer Downloaded!`nProcessing installer data. . . "
-      if ($_NewInstaller['InstallerType'] -eq 'msi') {
+      if ($_NewInstaller['InstallerType'] -in @('msi'; 'exe')) {
         $DetectedType = Get-PathInstallerType $script:dest
-        if ($DetectedType -in @('msi'; 'wix')) { $_NewInstaller['InstallerType'] = $DetectedType }
+        if ($DetectedType -in @('msi'; 'wix'; 'nullsoft'; 'inno')) { $_NewInstaller['InstallerType'] = $DetectedType }
```
Comment (mdanish-kh) [inline Tools/YamlCreate.ps1:1289]:
```suggestion
        if ($DetectedType -in @('msi'; 'wix'; 'nullsoft'; 'inno'; 'burn')) { $_NewInstaller['InstallerType'] = $DetectedType }
```

**Comment by russellbanks** [comment]
Scanning through the entire binary for a burn header seems a bit excessive, especially if the binary is big. Android Studio for example is > 1gb

**Comment by Trenly** [comment]
> Scanning through the entire binary for a burn header seems a bit excessive, especially if the binary is big. Android Studio for example is > 1gb

That's true, but it returns immediately as soon as it finds it if it is present; I'll have to see how long it takes to read a larger file, but because it's a streamreader it should be fairly quick

**Comment by Trenly** [comment]
> > Scanning through the entire binary for a burn header seems a bit excessive, especially if the binary is big. Android Studio for example is > 1gb
> 
> That's true, but it returns immediately as soon as it finds it if it is present; I'll have to see how long it takes to read a larger file, but because it's a streamreader it should be fairly quick. Its also why the NSIS/Inno checks are first

@russellbanks - Is there a specified number of bytes that komac will read?

**Comment by russellbanks** [comment]
> @russellbanks - Is there a specified number of bytes that komac will read?

Here's how it works simply:
- Offset starts at 0

Repeat the below 65535 times:
1. Read 8 (length of `.wixburn`) bytes at current offset
2. Compare those read bytes to `.wixburn` as bytes
3. If found, it's burn
3. Otherwise, move offset to the current position

If it's not found after 65535 times, it's not burn. This is the most performant way of doing it that I know of (takes milliseconds) and works every time (except for the JPackager case I mentioned in https://github.com/Trenly/winget-pkgs/issues/179).

Checking Inno and Nullsoft first is best priority-wise, as you're currently doing.

**Comment by russellbanks** [comment]
> So at most, 8*65535 bytes

No, I'm clearing the buffer so we only have 8 bytes in memory for reading at a time

**Comment by Trenly** [comment]
> > So at most, 8*65535 bytes
> 
> No, the buffer gets overwritten so we only have 8 bytes in memory for reading at a time

Right; But at most your read 8 bytes 65535 times, which would be 524280 bytes into the binary; If it isn't found within the first 524280 bytes, then it isn't burn

**Comment by russellbanks** [comment]
> > > So at most, 8*65535 bytes
> > 
> > 
> > No, the buffer gets overwritten so we only have 8 bytes in memory for reading at a time
> 
> Right; But at most your read 8 bytes 65535 times, which would be 524280 bytes into the binary; If it isn't found within the first 524280 bytes, then it isn't burn

Yeah

**Comment by Trenly** [comment]
@russellbanks - Latest commit limits it to just that maximum number of bytes; Please review

**Review by russellbanks — APPROVED** [review]

**Review by denelon — APPROVED** [review]

**Pull request merged**